arduino-core/src/cc/arduino/packages/discoverers/PluggableDiscovery.java

```diff
@@ -0,0 +1,213 @@
+import java.util.LinkedList;
+import java.util.List;
+import java.io.InputStream;
+import java.io.OutputStream;
+
+import cc.arduino.packages.BoardPort;
+import cc.arduino.packages.Discovery;
+import processing.app.legacy.PApplet;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.DeserializationFeature;
+import com.fasterxml.jackson.core.JsonFactory;
+import com.fasterxml.jackson.core.JsonParser;
+
+public class PluggableDiscovery implements Discovery {
+
+  private final String discoveryName;
+  private final String[] cmd;
+  private final List<BoardPort> portList;
+  private Process program=null;
+  private Thread pollingThread;
+
+  public PluggableDiscovery(String discoveryName, String[] cmd) {
+    this.cmd = cmd;
+    this.discoveryName = discoveryName;
+    portList = new LinkedList<>();
+    System.out.println(discoveryName + ": Starting: " + PApplet.join(cmd, " "));
+  }
+
+  @Override
+  public void run() {
+    // this method is started as a new thread, it will constantly listen
+    // to the discovery tool and keep track of the discovered ports
+    try {
+      start();
+      InputStream input = program.getInputStream();
+      JsonFactory factory = new JsonFactory();
+      JsonParser parser = factory.createParser(input);
+      ObjectMapper mapper = new ObjectMapper();
+      mapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
+
+      while (program != null && program.isAlive()) {
+        BoardPort port = mapper.readValue(parser, BoardPort.class);
+        if (port != null) {
+          System.out.println(discoveryName + ": received json");
+          String address = port.getAddress();
+          if (address != null) {
+            if (address.equals("Error: START_SYNC not supported")) {
+              if (pollingThread == null) {
+                startPolling();
+              }
+            } else {
+              update(port);
+            }
+          }
+        }
+      }
+      System.out.println(discoveryName + ": thread exit normally");
+    } catch (InterruptedException e) {
+      System.out.println(discoveryName + ": thread exit by interrupt");
+      e.printStackTrace();
+    } catch (Exception e) {
+      System.out.println(discoveryName + ": thread exit other exception");
+      e.printStackTrace();
+    }
+    try {
+      stop();
+    } catch (Exception e) {
+    }
+  }
+
+  @Override
+  public void start() throws Exception {
+    System.out.println(discoveryName + ": start");
+    try {
+      program = Runtime.getRuntime().exec(cmd);
+    } catch (Exception e) {
+      program = null;
+      return;
+    }
+    write("START_SYNC\n");
+    pollingThread = null;
+  }
+
+  private void startPolling() {
+    // Discovery tools not supporting START_SYNC require a periodic
+    // LIST command.  A second thread is created to send these
+    // commands, while the run() thread above listens for the
+    // discovery tool output.
```

Comment:
Not sure if this is the best way: we can call "LIST" just once directly in `listDiscoveredBoards()` (basically it will poll only when the Tools menu is triggered and not every 2.5 sec). On the other hand this may add some delay on opening the menu unless there is some internal cache in the discovery tool.

Reply:
We already do have the internal cache as the portList of discovered ports in PluggableDiscovery.  listDiscoveredBoards() just returns a copy of the list.  At least that part is easy.

To make this work, I believe DiscoverManager would need to give all discoverer instances an early notice that listDiscoveredBoards() may soon be called.  Maybe this could be done by detecting mouse hover over the Tools menu?  Or maybe mouse-over any part of the menus?

Early versions of Teensy's software waited in listDiscoveredBoards().  The result was usually ok, but some Windows & Mac users had a terrible experience with the GUI becoming unresponsive.  I believe it's best to always quickly return the list of previously discovered ports.